refactor: Add exception handling in background job within BOM Update Tool

marination opened the pull request:
> List View throws an error due to https://github.com/frappe/frappe/issues/16302 (its still usable)

**Issue:**
- BOM Update Tool jobs run in the background, there's no way to know about failures, if any (due to timeout or bug)
- This PR adds error logging and enhances the UX

**Working:**
- The **Replace** button is inactive until both BOMs are mentioned. On clicking on **Replace**, a **BOM Update Log** is created. This log maintains state and lets user know of failure.
   ![2022-03-17 13 01 58](https://user-images.githubusercontent.com/25857446/158759319-09cf13c7-98b1-4b3b-b0bb-8e0134aab7ac.gif)
- This log has two types: **Replace BOM** and **Update Cost**. Each is self explanatory as to what job is being executed.
- List View:
   <img width="1328" alt="Screenshot 2022-03-31 at 12 53 55 PM" src="https://user-images.githubusercontent.com/25857446/161000550-f607105c-bde7-4dd5-85d0-7cc55a5d5590.png">

- Failed Log:
  <img width="1337" alt="Screenshot 2022-03-17 at 1 05 44 PM" src="https://user-images.githubusercontent.com/25857446/158759557-9fe86be5-f01f-4c0d-9aab-2763a674e1dd.png">
- Completed Log
  <img width="1322" alt="Screenshot 2022-03-17 at 1 07 03 PM" src="https://user-images.githubusercontent.com/25857446/158759778-c4ad6579-66d4-486e-96b4-7f6912725fb3.png">
- Failed Log with simulated timeout error:
  ![2022-03-17 17 13 14](https://user-images.githubusercontent.com/25857446/158801598-95574fea-b8f4-421f-b54e-c85139d77cee.gif)

**Todo:**
- [x] Add BOM Update Log with status
- [x] Tests for Log

marination committed:
refactor: Add exception handling in background job within BOM Update Tool

marination committed:
Merge branch 'develop' into bom-update-tool

marination committed:
feat: BOM Update Log

- Created BOM Update Log that will handle queued job status and failures
- Moved validation and BG job to thus new doctype
- BOM Update Tool only works as an endpoint

marination committed:
chore: Polish error handling and code sepration

- Added Typing
- Moved all job business logic to bom update log
- Added `run_bom_job` that handles errors and runs either of two methods
- UX: Replace button disabled until both inputs are filled
- Show log creation message on UI for correctness
- APIs return log document as result
- Converted raw sql to QB

marination committed:
feat: List View indicators for Log and Error Log link in log

marination committed:
fix: Sider and Linter

marination committed:
fix: Test, Sider and Added button to access log from Tool

ankush committed:
Merge branch 'develop' into bom-update-tool

marination committed:
Merge branch 'develop' into bom-update-tool

marination committed:
test: API hit via BOM Update Tool

- test creation of log and it's impact

marination committed:
fix: Auto format `bom_update_log.py`

marination committed:
fix: Semgrep

- Explain explicit commits and skip semgrep
- Format client side translated string correctly

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -0,0 +1,163 @@
+	frappe.cache().delete_key("bom_children")
+	parent_boms = get_parent_boms(new_bom)
+
+	with click.progressbar(parent_boms) as parent_boms:
+		pass
```
Is this required? if yes, following should be part of contextmanager scope?

marination replied:
removing it,log is good enough for this cut. Job in normal cases shouldn't take that long also.

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -0,0 +1,163 @@
+	with click.progressbar(parent_boms) as parent_boms:
+		pass
+	for bom in parent_boms:
+		bom_obj = frappe.get_cached_doc("BOM", bom)
```
Since cache eviction isn't perfect, should we rely on the cached doc? 😅 

Any changes done via raw SQL / unsafe ORM methods might be lost.

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -0,0 +1,163 @@
+		bom_item.amount, (bom_item.stock_qty * unit_cost)
+	).where(
+		(bom_item.bom_no == current_bom) & (bom_item.docstatus < 2) & (bom_item.parenttype == "BOM")
+	).run()
```
This looks ugly (it is my fault though 🥲)


All select queries look better but updates are problematic since they aren't wrapped


Just wrapping it in `( )` makes it pretty:


 
<img width="984" alt="Screenshot 2022-03-30 at 3 24 43 PM" src="https://user-images.githubusercontent.com/9079960/160804635-83a2bd01-4179-4959-8f11-3eb0a4f1053f.png">

marination replied:
yeah it auto formatted this lol

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -0,0 +1,163 @@
+		# to make separate db calls by using load_doc_before_save
+		# which proves to be expensive while doing bulk replace
+		bom_obj._doc_before_save = bom_obj
+		bom_obj.update_new_bom(unit_cost, current_bom, new_bom)
```
BOM methodd has a different signature: https://github.com/frappe/erpnext/blob/ebf00946c91bf03105533d46c85e9b405cc7d62a/erpnext/manufacturing/doctype/bom/bom.py#L700 

perheps the new function added in this file is what you meant to call?

If yes, then this needs to be revisited, cause direct DB updates won't be present on BOM object from cache.

ankush replied:
also add a test for this, since currently build is passing _somehow_

marination replied:
I'll rename the new function.
meant to call the bom object function, must've gotten mixed up.
It passes because theres no checks and the snippet below always continues 😿 
<img width="307" alt="Screenshot 2022-03-30 at 4 28 31 PM" src="https://user-images.githubusercontent.com/25857446/160816424-0c10ec7f-2237-480c-aa11-835770f0577a.png">

marination replied:
@ankush Removed `update_new_bom()` in https://github.com/frappe/erpnext/pull/30146/commits/620575a9012a9759c6285558ac25c6709c4e92cc

Did the same thing `update_new_bom_in_bom_items()` does. Double check ?

ankush replied:
Yep, both seem to be doing same thing.

Actually no, update cost wasn't working since it's updating child items but not really saving to db 🤦 (db_update only updates main document not child items)

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -0,0 +1,163 @@
+
+		doc.db_set("status", "Completed")
+
+	except (Exception, JobTimeoutException):
```
```suggestion
	except Exception:
```

Second one inherits from `Exception` so it's same.

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -0,0 +1,163 @@
+
+
+def run_bom_job(
+	doc: "BOMUpdateLog", boms: Optional[Dict] = None, update_type: Optional[str] = "Replace BOM"
```
```suggestion
	doc: "BOMUpdateLog", boms: Optional[Dict[str, str]] = None, update_type: str = "Replace BOM"
```

- `boms` maps old name to new name, so encoding that info
- `update_type` isn't really optional, it just has a default value. For instance it can't be `None`? 

---

I see, the "Update Cost" isn't explicitly added when doing update cost. 


Type should be `Literal["Replace BOM", "Update Cost"]`

ankush commented on erpnext/manufacturing/doctype/bom_update_tool/bom_update_tool.py:
```diff
@@ -1,136 +1,67 @@
-	frappe.db.auto_commit_on_many_writes = 0
+
+def create_bom_update_log(
+	boms: Optional[Dict] = None, update_type: str = "Replace BOM"
```
why accept `None` as boms here? 🤔

marination replied:
In case of "Update Cost" type log, boms can be `None`, they dont need to be passed

marination committed:
fix: Type Annotations, Redundancy, etc.

- Renamed public function`update_new_bom` to `update_new_bom_in_bom_items`
- Replaced `get_cached_doc` with `get_doc`
- Removed click progress bar (drive through update log)
- Removed `bom_obj.update_new_bom()`, was redundant. Did same job as `update_new_bom_in_bom_items`
- Removed `update_new_bom()` in `bom.py`, unused.
- Prettier query formatting
- `update_type` annotated as non optional Literal
- Removed redundant use of JobTimeoutException
- Corrected type annotations in `create_bom_update_log()`

marination committed:
test: Added test for 2 more validations

- Covers full validate function

marination committed:
Merge branch 'develop' into bom-update-tool

marination commented:
@ankush as discussed will fix cost updation inconsistencies in a different PR. Good to go for now ?

marination committed:
chore: Added BOM std filters and update type in List View

ankush commented:
unrelated test failing.

marination commented:
Fixed the failing test in https://github.com/frappe/erpnext/pull/30509

marination committed:
Merge branch 'develop' into bom-update-tool
